### PR TITLE
Code running option

### PR DESCRIPTION
**Pull request by matebenyovszky**

### CODE REVIEW

File: aiAssistant.ts [hunk 1]
```diff
@@ -6,17 +6,18 @@ import { functions } from './functions'
...
 
 // Run this only once to create the assistant!
+// npm run create-ai-assistant
 export async function createAIDeveloper() {
   const aiDeveloper = await openai.beta.assistants.create({
     instructions: `You are an AI developer.
     When given a coding task, write and save code to files, install any packages if needed, make commits, and finally create a PR once done. You're logged in using GitHub CLI and have all the needed credentials.
     Start by listing all files inside the repo. You work inside the '/home/user/repo' directory where the repository is already cloned so you don't need to clone it yourself.
-    Don't argue with me and just complete the task.`,
+    Don't argue with me and just complete the task. You also have the ability to actually run the code or commands.`,
```
Comment (jakubno) [inline aiAssistant.ts]: You probably want something like this
```suggestion
    You also have the ability to actually run the Python code or Shell commands.
    Don't argue with me and just complete the task.`,
```

File: e2b.Dockerfile [hunk 2]
```diff
@@ -12,3 +12,7 @@ RUN curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | s
 	&& echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | sudo tee /etc/apt/sources.list.d/github-cli.list > /dev/null \
 	&& sudo apt update \
 	&& sudo apt install gh -y
+
+# Install Python
+RUN sudo apt update \
```
Comment (jakubno) [inline e2b.Dockerfile]: python is already installed in the template. Actually python should be in ubuntu by default

File: functions.ts [hunk 3]
```diff
@@ -98,4 +98,21 @@ export const functions: Array<
       },
     },
   },
+  // Run code
```
Comment (jakubno) [inline functions.ts:113]: this is probably too much general, either you want run just shell commands or python code, each should have different instructions. Otherwise the model can produce input which won't be compatible. E.g. generating just code for python without `python -c` before the code.

File: index.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ const gitName = 'e2b-assistant[bot]'
 
 const rootdir = '/home/user'
 const repoDir = 'repo'
-const repoDirPath = path.join(rootdir, repoDir)
+const repoDirPath = path.posix.join(rootdir, repoDir);
```
Comment (jakubno) [inline index.ts]:
```suggestion
const repoDirPath = path.posix.join(rootdir, repoDir)
```

File: index.ts [hunk 5]
```diff
@@ -67,6 +67,17 @@ async function cloneRepo(sandbox: Sandbox, repo: string) {
 	await setRemote.wait()
 }
 
+async function listLastFiveRepos(sandbox: Sandbox) {
```
Comment (jakubno) [inline index.ts]: this probably shouldn't be part of the PR

**Review by jakubno — CHANGES_REQUESTED** [review]
Could you also resolve the conflicts please

File: e2b.Dockerfile (hunk 2 above)
Comment (matebenyovszky) [reply e2b.Dockerfile]: if i remember correctly "python3" worked... but will test again

File: functions.ts (hunk 3 above)
Comment (matebenyovszky) [reply functions.ts:113]: it just worked for me, I was able to run python, github cli. even I am thinking on simplifing the original tools that you made, becuase it can just ran this like "run a github command to merge and rebase pull requests". Or make a github tool that does anything requested through the github cli

Comment (matebenyovszky) [reply functions.ts:113]: but probably i will iterate with this later

**Comment by jakubno** [comment]
- There are still some conflicts
- What's the purpose of listing last five repos? If you want help the user, use `select` from `prompts` package. But you probably want them to decide, if they want to write or choose. Also for choosing, you will probably want more than 5 and you'll have to add option to load next "page"

**Comment by matebenyovszky** [comment]
@jakubno sorry, I just try to learn coding and GitHub logic... and for sure, it is much harder than reviewing a document in Word - which is in issue in my opinion. Do we have an AI that can simplify these Github reviews? Sound like a good project to me :) :)
(If there are still issues with my pull request, please feel free to dismiss or correct that)

File: index.ts [hunk 6]
```diff
@@ -257,11 +253,34 @@ const assistant = await getAssistant()
 const sandbox = await Sandbox.create({ id: 'ai-developer-sandbox', onStdout: onLog, onStderr: onLog })
 await loginWithGH(sandbox)
 
-await listLastFiveRepos(sandbox)
-
 const { repoName, task } = await initChat()
 
 // Start terminal session with user
+=======
```
Comment (matebenyovszky) [inline index.ts]: this has to be removed

File: readme.md [hunk 7]
```diff
@@ -34,7 +36,7 @@ The custom E2B sandbox environment is defined in the [`e2b.Dockerfile`](./e2b.Do
 1. Add your OpenAI API key to `.env`
 1. Get E2B API Key at [https://e2b.dev/docs/getting-started/api-key](https://e2b.dev/docs/getting-started/api-key)
     - Save it to `.env`
-1. Run `npm run create-ai-assistant` to create the AI assistant using OpenAI's new Assistant API
+1. Run `npm run create-ai` to create the AI assistant using OpenAI's new Assistant API
```
Comment (jakubno) [inline readme.md]: I would keep the old name, `create-ai` can be confusing, same for update

File: index.ts [hunk 8]
```diff
@@ -9,32 +9,66 @@ import { Sandbox } from '@e2b/sdk'
...
 
-  return { repoName, task }
+  return { repoName, task, mode }
+}  
+
+function createThread(repoURL: string, task: string, mode: string) {
+  let content;
+  if (mode === 'repo') {
+    content = `Pull this repo: '${repoURL}'. Then carefully plan this task and start working on it: ${task}`;
+  } else if (mode === 'chit-chat') {
```
Comment (jakubno) [inline index.ts]: What's this good for? I would prefer to keep this simple and on the point, then introducing feature which won't be used

File: actions.ts [hunk 9]
```diff
@@ -91,3 +91,52 @@ export async function readFile(sandbox: Sandbox, { path }: { path: string }): Pr
...
+      // Replace all non-ASCII characters in the output
+      const cleanedOutput = output.stdout.replace(/[^\x00-\x7F]/g, '')
+
+      return cleanedOutput
+  } catch (e) {
+      return `Error: ${e.message}`
+  }
+}
+
+export async function runPowershellCommand(sandbox: Sandbox, { command }: { command: string }): Promise<string> {
```
Comment (jakubno) [inline actions.ts]: I would remove this for multiple reasons:
- not that relevant to the goal of the app. It works in linux system. 
- the functionality can be achieved by `runCode` and I believe we discussed something similar for `python` and the outcome was the `runCode` is enough
- Pwsh isn't installed there by default, we can add, no problem with that, but imho shell is just enough :))

File: aiAssistant-update.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import OpenAI from 'openai'
+import 'dotenv/config'
+
+import { functions } from './functions'
+import { assistant_config } from './functions';
+
+const openai = new OpenAI()
+
+// Run this to update the assistant!
+// npm run create-ai
```
Comment (jakubno) [inline aiAssistant-update.ts]:
```suggestion
// npm run update-ai-assistant
```

File: functions.ts [hunk 11]
```diff
@@ -1,5 +1,17 @@
 import { AssistantCreateParams } from 'openai/src/resources/beta/assistants/assistants'
 
+export const assistant_config =
```
Comment (jakubno) [inline functions.ts:3]: not sure if this is a correct file for the config

**Review by jakubno — CHANGES_REQUESTED** [review]
I really like the update assistant and extracting the config! Fix the few things I mentioned and let's get it merged 🚀

File: readme.md (hunk 7 above)
Comment (matebenyovszky) [reply readme.md]: ok, apdated both in package.json too

File: actions.ts (hunk 9 above)
Comment (matebenyovszky) [reply actions.ts]: My goal was to be able to develop and run fairly complex PowerShel scripts, but I can accept that this is not needed for you

File: index.ts (hunk 8 above)
Comment (matebenyovszky) [reply index.ts]: ok

**Comment by matebenyovszky** [comment]
> not sure if this is a correct file for the config 

I found this the closest, but of could be anything else

**Comment by matebenyovszky** [comment]
Thanks for the review, I hope I updated everything

**Comment by matebenyovszky** [comment]
@jakubno please let me know if I could do anything else, I think I've finished with all changes

**Comment by ValentaTomas** [comment]
Hey @matebenyovszky I did some partial cleanup on https://github.com/e2b-dev/ai-developer-with-sandbox/tree/base branch a few days ago, maybe we can integrate that to your PR too?

**Comment by matebenyovszky** [comment]
Absolutely, I am new to Github, so what would be the strategy? My PR is nearly 2 weeks old, so if you like that I would be happy if it could be merged first quickly. But of course no pressure to accept it.
My main fork has some other changes which I removed from the PR upon your request, so 🤔... I would definitely appreciate if you can sort that out 🙄

**Comment by matebenyovszky** [comment]
Anyway, what would be the process @ValentaTomas ?

**Comment by ValentaTomas** [comment]
Got it. I think merging your PR first sounds like the best approach. Let's merge it when @jakubno deems this PR to be ready. Then we can check which changes on my branch are still relevant and potentially create a new PR.

File: actions.ts [hunk 12]
```diff
@@ -91,3 +91,27 @@ export async function readFile(sandbox: Sandbox, { path }: { path: string }): Pr
...
+      const process = await sandbox.process.start({ cmd: command, cwd: repoDirPath })
+      const output = await process.wait()
+
+      if (output.exitCode !== 0) {
+          throw new Error(`Command failed with exit code ${output.exitCode}. Error: ${output.stderr}`)
+      }
+
+      if (!output.stdout) {
+          throw new Error(`Command did not produce any output. Error: ${output.stderr}`)
+      }
```
Comment (jakubno) [inline actions.ts]: This doesn't need to be necessarily bad. E.g `touch`, `mkdir`, `rm` and many more won't produce any output
```suggestion
          console.log(`Command did not produce any output. Error: ${output.stderr}`)
      }
```

File: aiFunctions.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
+import { AssistantCreateParams } from 'openai/src/resources/beta/assistants/assistants'
+
+export const assistant_config = 
+{
+  instructions: `Depemding on the task, you are an AI assistant how can run bash, python and powershell codes in a terminal according to the task, or an AI developer. If You are an AI Developer:
```
Comment (jakubno) [inline aiFunctions.ts]: I don't understand this paragraph, could you please improve the phrasing, also remove the part with `powershell`

File: index.ts [hunk 14]
```diff
@@ -9,27 +9,54 @@ import { Sandbox } from '@e2b/sdk'
...
 import { sleep } from './sleep'
-import { listFiles, makeCommit, makeDir, makePullRequest, readFile, saveCodeToFile } from './actions'
+import { listFiles, makeCommit, makeDir, makePullRequest, readFile, saveCodeToFile, runCode } from './actions'
+
+import { listAIDeveloper } from './aiAssistant-list'
 
 const openai = new OpenAI()
 
-const AI_ASSISTANT_ID = process.env.AI_ASSISTANT_ID!
+//const AI_ASSISTANT_ID = process.env.AI_ASSISTANT_ID!
```
Comment (jakubno) [inline index.ts]:
```suggestion
```

**Review by jakubno — CHANGES_REQUESTED** [review]
Sorry for the delay, I finally found some time to get back. Can we solve these last issues and let's get it merged

**Review by matebenyovszky — COMMENTED** [review]
updated according to requests

File: index.ts (hunk 14 above)
Comment (jakubno) [inline index.ts]:
```suggestion
```

File: aiFunctions.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        type: 'object',
+        properties: {
+          command: {
+            type: 'string',
+            description: 'The PowerShell command to run',
+          },
+        },
+      },
+    },
+  },
```
Comment (jakubno) [inline aiFunctions.ts]:
```suggestion
```

**Comment by jakubno** [comment]
@matebenyovszky thanks for your contribution!
